modules/mesh-layers/src/utils/matrix.js

```diff
@@ -115,11 +115,11 @@ export const MATRIX_ATTRIBUTES = {
 
       const valueMatrix = new Float32Array(matrix);
       const valueTranslation = new Float32Array(modelTranslation);
-      const shaderAttributes = attribute.userData.shaderAttributes;
-      shaderAttributes.instanceModelMatrix__LOCATION_0.value = valueMatrix.subarray(0, 3);
-      shaderAttributes.instanceModelMatrix__LOCATION_1.value = valueMatrix.subarray(3, 6);
-      shaderAttributes.instanceModelMatrix__LOCATION_2.value = valueMatrix.subarray(6, 9);
-      shaderAttributes.instanceTranslation.value = valueTranslation;
+
+      attribute.userData.shaderValues.instanceModelMatrix__LOCATION_0 = valueMatrix.subarray(0, 3);
```

Comment:
You should be able to set `attribute.value` to an array of 16 here and derive the shader attribute values using their offsets?

Reply:
Clarify: in `attribute.update`, if a constant value is provided and the attribute accessor contains `offset`, should it automatically extract the correct subarray from the value?

Reply:
`offset` is attribute pointer state and using a constant attribute involves disabling the attribute pointer.

Reply:
But you can use the information to extract the subarray from the value in https://github.com/uber/deck.gl/blob/master/modules/core/src/lib/base-attribute.js#L63, something like

```
if (constant) {
	this.value = value.slice(this.offset / value.BYTES_PER_ELEMENT, this.size);
}
```

Reply:
As discussed, I don't think users of attributes should be accessing/manipulating arbitrary content in the `userData` field. Eventually, shader attributes handling should be strictly internal to the `Attribute` class.

Reply:
Oh yeah, that should work. Is value always a typed array?

Reply:
Hmm no